src/storage/src/hummock/iterator/merge_inner.rs

```diff
@@ -183,93 +188,100 @@ where
 /// The behaviour of `next` of order aware merge iterator is different from the normal one, so we
 /// extract this trait.
 trait MergeIteratorNext {
-    type HummockResultFuture<'a>: Future<Output = HummockResult<()>> + Send + 'a
-    where
-        Self: 'a;
-    fn next_inner(&mut self) -> Self::HummockResultFuture<'_>;
+    fn poll_next_inner(&mut self) -> Poll<HummockResult<()>>;
 }
 
 impl<I: HummockIterator> MergeIteratorNext for OrderedMergeIteratorInner<I> {
-    type HummockResultFuture<'a> = impl Future<Output = HummockResult<()>>;
-
-    fn next_inner(&mut self) -> Self::HummockResultFuture<'_> {
-        async {
-            let top_node = self.heap.pop().expect("no inner iter");
-            let mut popped_nodes = vec![];
-
-            // Take all nodes with the same current key as the top_node out of the heap.
-            while let Some(next_node) = self.heap.peek_mut() {
-                match VersionedComparator::compare_key(top_node.iter.key(), next_node.iter.key()) {
-                    Ordering::Equal => {
-                        popped_nodes.push(PeekMut::pop(next_node));
-                    }
-                    _ => break,
+    fn poll_next_inner(&mut self) -> Poll<HummockResult<()>> {
+        debug_assert!(self.pending_iters.is_empty());
+        let top_node = self.heap.pop().expect("no inner iter");
+        let mut popped_nodes = vec![];
+
+        // Take all nodes with the same current key as the top_node out of the heap.
+        while let Some(next_node) = self.heap.peek_mut() {
+            match VersionedComparator::compare_key(top_node.iter.key(), next_node.iter.key()) {
+                Ordering::Equal => {
+                    popped_nodes.push(PeekMut::pop(next_node));
                 }
+                _ => break,
             }
+        }
 
-            popped_nodes.push(top_node);
+        popped_nodes.push(top_node);
 
-            // WARNING: within scope of BinaryHeap::PeekMut, we must carefully handle all places of
-            // return. Once the iterator enters an invalid state, we should remove it from heap
-            // before returning.
+        // WARNING: within scope of BinaryHeap::PeekMut, we must carefully handle all places of
+        // return. Once the iterator enters an invalid state, we should remove it from heap
+        // before returning.
 
-            // Put the popped nodes back to the heap if valid or unused_iters if invalid.
-            for mut node in popped_nodes {
-                match node.iter.next().await {
+        // Put the popped nodes back to the heap if valid or unused_iters if invalid.
+        for mut node in popped_nodes {
```

Comment:
This state machine doesn't look equivalent to the previous one. For this one, if there's any iterator pending, we will go over the `while let Some(next_node) = self.heap.peek_mut()` loop when calling `poll_next_inner`.

Reply:
Before we call this method we should ensure that there is no pending iterator. We should call `await_next` after Poll::Pending is returned from `poll_next`, and the `await_next` should have awaited all pending iterators.

The usage of `poll_next` is different to the conventional usage of `poll` in that when `Pending` is returned, we should not call `poll_next` again, and should call `await_next`.